crates/wasm-compose/src/graph.rs

```diff
@@ -737,17 +911,38 @@ impl<'a> CompositionGraph<'a> {
                 .export_entity_type(export_index)
                 .ok_or_else(|| anyhow!("the source export index is invalid"))?;
 
-            if !ComponentEntityType::is_subtype_of(
-                &export_ty,
+            let resource_mapping = self.resource_mapping.borrow().clone();
+
+            if let Some(resource_mapping) = resource_mapping.add_pairs(
+                source_instance.component,
+                export_ty,
                 source_component.types(),
-                &import_ty,
+                target_instance.component,
+                import_ty,
                 target_component.types(),
             ) {
-                bail!(
-                    "source {export_ty} export `{export_name}` is not compatible with target {import_ty} import `{import_name}`",
-                    export_ty = type_desc(export_ty),
-                    import_ty = type_desc(import_ty),
-                );
+                let remapping = &mut resource_mapping.remapping();
+                let mut context =
+                    SubtypeCx::new_with_refs(source_component.types(), target_component.types());
+
+                let mut a_ty = export_ty;
+                context.a.remap_component_entity(&mut a_ty, remapping);
+                remapping.reset();
+
+                let mut b_ty = import_ty;
+                context.b.remap_component_entity(&mut b_ty, remapping);
+                remapping.reset();
+
+                if context.component_entity_type(&a_ty, &b_ty, 0).is_err() {
```

Comment:
This same sequence of code shows up in 2 other places I've seen so far now, could it instead be refactored into a shared helper used by this crate?